sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/ApproximateDistinct.java

```diff
@@ -0,0 +1,600 @@
+
+  /**
+   * The {@link PTransform} takes an input {@link PCollection} of objects and returns a
+   * {@link PCollection} whose contents is a sketch that can be queried in order to retrieve
+   * the approximate number of distinct elements in the input {@link PCollection}.
+   *
+   * @param <InputT>    the type of the elements in the input {@link PCollection}
+   */
+  public static <InputT> GloballyDistinct<InputT> globally() {
+    return GloballyDistinct.<InputT>builder()
+            .build();
+  }
+
+  /**
+   * A {@link PTransform} that takes an input {@code PCollection<KV<K, V>}
+   * and returns a {@code PCollection<KV<K, HyperLogLogPlus>>} that contains an
+   * output element mapping each distinct key in the input {@link PCollection}
+   * to a structure wrapping a {@link HyperLogLogPlus}. It can be queried in
+   * order to retrieve the approximate number of distinct values associated with
+   * each key in the input {@link PCollection}.
+   *
+   * @param <K>         type of the keys mapping the elements
+   * @param <V>         type of the values being combined per key
+   */
+  public static <K, V> PerKeyDistinct<K, V> perKey() {
+    return PerKeyDistinct.<K, V>builder()
+            .build();
+  }
+
+  /**
+   * Implementation of {@link #globally()}.
+   *
+   * @param <InputT>    the type of the elements in the input {@link PCollection}
+   */
+  @AutoValue
+  public abstract static class GloballyDistinct<InputT>
+          extends PTransform<PCollection<InputT>, PCollection<Long>> {
+
+    @Nullable
+    abstract Coder<InputT> inputCoder();
+
+    abstract int precision();
+
+    abstract int sparsePrecision();
+
+    abstract Builder<InputT> toBuilder();
+
+    static <InputT> Builder<InputT> builder() {
+      return new AutoValue_ApproximateDistinct_GloballyDistinct.Builder<InputT>()
+              .setInputCoder(null)
+              .setPrecision(12)
+              .setSparsePrecision(0);
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder<InputT> {
+      abstract Builder<InputT> setInputCoder(Coder<InputT> coder);
+
+      abstract Builder<InputT> setPrecision(int p);
+
+      abstract Builder<InputT> setSparsePrecision(int sp);
+
+      abstract GloballyDistinct<InputT> build();
+    }
+
+
+    public GloballyDistinct<InputT> withInputCoder(Coder<InputT> coder) {
+      return toBuilder().setInputCoder(coder).build();
+    }
+
+    public GloballyDistinct<InputT> withPrecision(int p) {
+      return toBuilder().setPrecision(p).build();
+    }
+
+    public GloballyDistinct<InputT> withSparsePrecision(int sp) {
+      return toBuilder().setSparsePrecision(sp).build();
+    }
+
+    @Override
+    public PCollection<Long> expand(PCollection<InputT> input) {
+      return input
+              .apply("Compute HyperLogLog Structure", Combine.globally(ApproximateDistinctFn
+                      .<InputT>create(MoreObjects.firstNonNull(this.inputCoder(), input.getCoder()))
```

Comment:
Not sure why inputCoder() is needed at all? The input collection always has a coder, so can we just always use it?

Reply:
Yes you're right it seems that in most cases the users won't specify a different coder for the input. I just misunderstood one of Luke's comments.